src/lvl_script.c

```diff
@@ -4119,6 +4214,162 @@ TbBool script_level_up_creature(PlayerNumber plyr_idx, long crmodel, long criter
+        break;
+    case CSelCrit_LeastExpFighting:
+        thing = find_players_lowest_level_creature_of_breed_and_gui_job(crmodel, CrGUIJob_Fighting, plyr_idx, 0);
+        break;
+    case CSelCrit_NearOwnHeart:
+    {
+        const struct Coord3d* pos = dungeon_get_essential_pos(plyr_idx);
+        thing = get_creature_near_and_owned_by(pos->x.val, pos->y.val, plyr_idx);
+        break;
+    }
+    case CSelCrit_NearEnemyHeart:
+        thing = get_creature_in_range_around_any_of_enemy_heart(plyr_idx, crmodel, 11);
+        break;
+    case CSelCrit_OnEnemyGround:
+        thing = get_random_players_creature_of_model_on_territory(plyr_idx, crmodel,0);
+        break;
+    case CSelCrit_OnFriendlyGround:
+        thing = get_random_players_creature_of_model_on_territory(plyr_idx, crmodel,1);
+        break;
+    default:
+        ERRORLOG("Invalid level up criteria %d",(int)criteria);
+        thing = INVALID_THING;
+        break;
+    }
+    if (thing_is_invalid(thing)) {
+        SYNCDBG(5,"No matching player %d creature of model %d found to cast spell on.",(int)plyr_idx,(int)crmodel);
+        return false;
+    }
+    
+    // get spell type from left octet
+    PowerKind powerKind = magic_caster_splvl >> 16;
+    // get caster PlayerNumber from middle octet
+    PlayerNumber caster =  (magic_caster_splvl >> 8) & ((1 << 8) - 1);
+    // get spell level from right octet
+    long splevel = (magic_caster_splvl & ((1 << 8) - 1));
+    MapSubtlCoord thingXPos = thing->mappos.x.stl.num;
+    MapSubtlCoord thingYPos = thing->mappos.y.stl.num;
+    switch (powerKind)
+    {
+        case PwrK_DISEASE:
+            return magic_use_power_disease(caster, thing, 0, 0, splevel, PwMod_CastForFree);
+        case PwrK_SPEEDCRTR:
+            return magic_use_power_speed(caster, thing, 0, 0, splevel, PwMod_CastForFree);
+        case PwrK_SLAP:
+            return magic_use_power_slap_thing(caster, thing, PwMod_CastForFree);
+        case PwrK_CHICKEN:
+            return magic_use_power_chicken(caster, thing, 0, 0, splevel, PwMod_CastForFree);
+        case PwrK_CONCEAL:
+            return magic_use_power_conceal(caster, thing, 0, 0, splevel, PwMod_CastForFree);
+        case PwrK_HEALCRTR:
+            return magic_use_power_heal(caster, thing, 0, 0, splevel, PwMod_CastForFree);
+        case PwrK_PROTECT:
+            return magic_use_power_armour(caster, thing, 0, 0, splevel, PwMod_CastForFree);
+        case PwrK_LIGHTNING:
+            return magic_use_power_lightning(caster, thingXPos, thingYPos, splevel, PwMod_CastForFree);
+        case PwrK_CAVEIN:
+            return magic_use_power_cave_in(caster, thingXPos, thingYPos, splevel, PwMod_CastForFree);
+        case PwrK_CALL2ARMS:
+            return magic_use_power_call_to_arms(caster, thingXPos, thingYPos, splevel, PwMod_CastForFree);
+        case PwrK_MKDIGGER:
+            return magic_use_power_imp(caster, thingXPos, thingYPos, PwMod_CastForFree);
+        default:
```

Comment:
You may use `magic_use_power_on_thing` instead of this big switch

Reply:
He mentioned it below, but he included spells that are not cast on things. This way the mapmaker can cast for example the sight spell on a specific enemy unit, revealing it.